compiler/src/dotty/tools/dotc/printing/PlainPrinter.scala

```diff
@@ -731,7 +729,57 @@ class PlainPrinter(_ctx: Context) extends Printer {
     ~ (if param.isTypeParam then "" else ": ")
     ~ toText(param.paramInfo)
 
-  protected def escapedString(str: String): String = str flatMap escapedChar
+  protected final def escapedString(str: String): String = escapedString(str, quoted = false)
+
+  private def requiresFormat(c: Char): Boolean = (c: @switch) match
+    case '\b' | '\t' | '\n' | '\f' | '\r' | '"' | '\'' | '\\' => true
+    case c => c.isControl
+
+  private def escapedString(text: String, quoted: Boolean): String =
+    def mustBuild: Boolean =
+      var i = 0
+      while i < text.length do
+        if requiresFormat(text.charAt(i)) then return true
+        i += 1
+      false
+    if mustBuild then
+      val b = StringBuilder(text.length + 16)
+      if quoted then
+        b.append('"')
+      var i = 0
+      while i < text.length do
+        escapedChar(b, text.charAt(i))
+        i += 1
+      if quoted then
+        b.append('"')
+      b.toString
+    else if quoted then "\"" + text + "\""
+    else text
+
+  private def escapedChar(b: StringBuilder, c: Char): Unit =
```

Comment:
Can we move the logic to `compiler/src/dotty/tools/dotc/util/Chars.scala` to avoid duplication?

Reply:
I guess we can!